File: apps/csharp/tests/Ch13/WordCountTests.cs
using GrokkingConcurrency.CSharp.Ch13;

namespace GrokkingConcurrency.CSharp.Tests.Ch13;

/// <summary>
/// MapReduce パターンによるワードカウントのテスト
/// </summary>
public class WordCountTests
{
    [Fact]
    public void ShouldMapTextToWordCountPairs()
    {
        var text = "hello world hello";

        var pairs = WordCount.Map(text);

        Assert.Equal(3, pairs.Count);
        Assert.Contains(pairs, p => p.Key == "hello" && p.Value == 1);
        Assert.Contains(pairs, p => p.Key == "world" && p.Value == 1);
    }

    [Fact]
    public void ShouldReduceWordCountPairs()
    {
        var pairs = new List<KeyValuePair<string, int>>
        {
            new("hello", 1),
            new("world", 1),
            new("hello", 1)
        };

        var result = WordCount.Reduce(pairs);

        Assert.Equal(2, result["hello"]);
        Assert.Equal(1, result["world"]);
    }

    [Fact]
    public void ShouldCountWordsInMultipleTexts()
    {
        var texts = new List<string>
        {
            "hello world",
            "hello again",
            "world peace"
        };

        var result = WordCount.CountWords(texts);

        Assert.Equal(2, result["hello"]);
        Assert.Equal(2, result["world"]);
        Assert.Equal(1, result["again"]);
        Assert.Equal(1, result["peace"]);
    }
}

File: apps/fsharp/tests/WordCountTests.fs
module WordCountTests

open Xunit
open GrokkingConcurrency.Ch13.WordCount

[<Fact>]
let ``map converts text to word-count pairs`` () =
    let result = map "hello world hello"
    Assert.Contains(("hello", 1), result)
    Assert.Contains(("world", 1), result)
    let helloCount = result |> List.filter (fun (w, _) -> w = "hello") |> List.length
    Assert.Equal(2, helloCount)

[<Fact>]
let ``reduce aggregates word counts`` () =
    let pairs = [("hello", 1); ("world", 1); ("hello", 1)]
    let result = reduce pairs
    Assert.Equal(2, Map.find "hello" result)
    Assert.Equal(1, Map.find "world" result)

[<Fact>]
let ``countWords performs MapReduce`` () =
    let texts = [
        "hello world"
        "hello fsharp"
        "world of fsharp"
    ]
    let result = countWords texts
    Assert.Equal(2, Map.find "hello" result)
    Assert.Equal(2, Map.find "world" result)
    Assert.Equal(2, Map.find "fsharp" result)
    Assert.Equal(1, Map.find "of" result)
